ergm.py

```diff
@@ -187,6 +201,127 @@ def sample_network(self, sampling_method="NaiveMetropolisHastings", seed_network
 
         return network
 
+
+class BruteForceERGM(ERGM):
+    """
+    A class that implements ERGM by iterating over the entire space of networks and calculating stuff exactly (rather
+    than using statistical methods for approximating and sampling).
+    This is mainly for tests.
+    """
+    # The maximum number of nodes that is allowed for carrying brute force calculations (i.e. iterating the whole space
+    # of networks and calculating stuff exactly). This becomes not tractable above this limit.
+    MAX_NODES_BRUTE_FORCE = 6
+
+    @staticmethod
+    def construct_adj_mat_from_int(int_code: int, num_nodes: int, is_directed: bool) -> np.ndarray:
+        num_pos_connects = num_nodes * (num_nodes - 1)
+        if not is_directed:
+            num_pos_connects //= 2
+        adj_mat_str = f'{int_code:0{num_pos_connects}b}'
+        cur_adj_mat = np.zeros((num_nodes, num_nodes))
+        mat_entries_arr = np.array(list(adj_mat_str), 'uint8')
+        if is_directed:
+            cur_adj_mat[~np.eye(num_nodes, dtype=bool)] = mat_entries_arr
+        else:
+            upper_triangle_indices = np.triu_indices(num_nodes, k=1)
+            cur_adj_mat[upper_triangle_indices] = mat_entries_arr
+            lower_triangle_indices_aligned = (upper_triangle_indices[1], upper_triangle_indices[0])
+            cur_adj_mat[lower_triangle_indices_aligned] = mat_entries_arr
+        return cur_adj_mat
+
+    @staticmethod
+    def construct_int_from_adj_mat(adj_mat: np.ndarray, is_directed: bool) -> int:
+        if len(adj_mat.shape) != 2 or adj_mat.shape[0] != adj_mat.shape[1]:
+            raise ValueError(f"The dimensions of the given adjacency matrix {adj_mat.shape} are not valid for an "
+                             f"adjacency matrix (should be a 2D squared matrix)")
+        num_nodes = adj_mat.shape[0]
+        adj_mat_no_diag = adj_mat[~np.eye(num_nodes, dtype=bool)]
+        if not is_directed:
+            adj_mat_no_diag = adj_mat_no_diag[:adj_mat_no_diag.size // 2]
+        return round((adj_mat_no_diag * 2 ** np.arange(adj_mat_no_diag.size - 1, -1, -1).astype(np.ulonglong)).sum())
+
+    def __init__(self,
+                 n_nodes,
+                 network_statistics: NetworkStatistics,
+                 is_directed=False,
+                 initial_thetas=None,
+                 initial_normalization_factor=None,
+                 seed_MCMC_proba=0.25):
+        super().__init__(n_nodes,
+                         network_statistics,
+                         is_directed,
+                         initial_thetas,
+                         initial_normalization_factor,
+                         seed_MCMC_proba)
+        self._all_weights = self._calc_all_weights()
+        self._normalization_factor = self._all_weights.sum()
+
+    # TODO: think of an alternative name? confusing with the static method
```

Comment:
Why do we even need this function? instead of just calling `construct_adj_mat_from_int` directly?

Reply:
(which you do with `construct_int_from_adj_mat` in `calculate_weight`)